src/components/puck/registry/build-registry.ts

```diff
@@ -0,0 +1,101 @@
+// Based on https://github.com/shadcn-ui/ui/blob/main/apps/www/scripts/build-registry.mts
+import { registryComponents } from "./registry.ts";
+import { promises as fs, existsSync } from "fs";
+import { z } from "zod";
+import { registryItemFileSchema } from "./schema.ts";
+import path from "path";
+
+const REGISTRY_BASE_PATH = "./src/components/puck";
+const PUBLIC_FOLDER_BASE_PATH = "./src/components/puck/registry/output";
+const PUBLIC_FOLDER_INDEX_PATH = `${PUBLIC_FOLDER_BASE_PATH}/index.ts`;
+const COMPONENT_FOLDER_PATH = "components";
+const VISUAL_EDITOR_PATH = "../../index.ts";
+
+type File = z.infer<typeof registryItemFileSchema>;
+
+async function writeFileRecursive(filePath: string, data: string) {
+  const dir = path.dirname(filePath); // Extract the directory path
+
+  try {
+    // Ensure the directory exists, recursively creating directories as needed
+    await fs.mkdir(dir, { recursive: true });
+
+    // Write the file
+    await fs.writeFile(filePath, data, "utf-8");
+  } catch (error) {
+    console.error("Error writing file: ", error);
+  }
+}
+
+const getComponentFiles = async (files: File[]) => {
+  const filesArrayPromises = (files ?? []).map(async (file) => {
+    if (typeof file === "string") {
+      const filePath = `${REGISTRY_BASE_PATH}/${file}`;
+      const fileContent = await fs.readFile(filePath, "utf-8");
+      return {
+        type: "registry:component",
+        content: fileContent.replaceAll(
+          VISUAL_EDITOR_PATH,
+          "@yext/visual-editor"
+        ),
+        path: file,
+        target: `${COMPONENT_FOLDER_PATH}/${file}`,
+      };
+    }
+  });
+  const filesArray = await Promise.all(filesArrayPromises);
+
+  return filesArray;
+};
+
+const main = async () => {
+  // Delete index.ts if it exists
+  if (existsSync(PUBLIC_FOLDER_INDEX_PATH)) {
+    await fs.rm(PUBLIC_FOLDER_INDEX_PATH);
+  }
+
+  const componentNames = [];
+
+  // make a json file and put it in public folder
+  for (let i = 0; i < registryComponents.length; i++) {
+    const component = registryComponents[i];
+    componentNames.push(component.name);
+    const files = component.files;
+    if (!files) throw new Error("No files found for component");
+
+    const filesArray = await getComponentFiles(files);
+
+    const json = JSON.stringify(
+      {
+        ...component,
+        files: filesArray,
+      },
+      null,
+      2
+    );
+    const jsonPath = `${PUBLIC_FOLDER_BASE_PATH}/${component.name}.json`;
+    await writeFileRecursive(jsonPath, json);
+
+    // import new json file in index.ts
+    await fs.appendFile(
+      PUBLIC_FOLDER_INDEX_PATH,
+      `import { default as ${component.name} } from "./${component.name}.json"\n`,
```

Comment:
Is there a way to to make it work like:
`npx shadcn add https://reliably-numerous-kit.pgsdemo.com/components/fullList.json HeadingText`

I see in their example it's what you have, but it would be nice to point to a single entrypoint and name the component, rather than a different url per component. I assume the answer is no. It's worth seeing if there are any issues for something like that and if not making one.

Reply:
Switched around the serving repo so now
`REGISTRY_URL=https://reliably-numerous-kit.pgsdemo.com/components npx shadcn add HeadingText` works

Reply:
Awesome 🔥

Reply:
What defines the url path or filename now?

Reply:
It's `REGISTRY_URL/styles/STYLE/NAME.json` with NAME being what you put after add and STYLE being read from  the `components.json` file in the destination repo (this is the structure of pulling normal shadcn components). I think I need to TAL a little more at how much pulling our components would block you from pulling shadcn components and vice versa

Reply:
Well pulling shadcn components you just wouldn't defined a registry-url, right? So you're saying NAME.json is coming from the fact that the file is named components.json? Was just looking to know how components gets set as the filename for https://reliably-numerous-kit.pgsdemo.com/components.

And the domain doesn't matter in this package then now, right? Because the user just puts in whatever the registry url is.

Reply:
Realizing now that you export the registry from this package and import it in another [repo](https://github.com/yext-pages/visual_editor_library_4174974/blob/master/index.js) which builds it. Let's have this repo do that directly instead. We won't need to export the registry anymore, but instead have a registry build cmd in package.json that generates the json files into `dist` and then have a config.yaml and artifacts.json. The build step would need to copy artifacts.json to dist as well.

Note that there's two separate things. The first is a normal build for the lib which goes to dist. But to host the registry files we'd have a separate build command which also goes to dist. Both build commands shouldn't ever need to be run at the same time since there are two separate purposes.